libs/x-content/src/main/java/org/elasticsearch/xcontent/DotExpandingXContentParser.java

```diff
@@ -8,15 +8,23 @@
 
 package org.elasticsearch.xcontent;
 
+import org.elasticsearch.core.CheckedFunction;
+
 import java.io.IOException;
 import java.util.ArrayDeque;
 import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
 
 /**
  * An XContentParser that reinterprets field names containing dots as an object structure.
  *
  * A field name named {@code "foo.bar.baz":...} will be parsed instead as {@code 'foo':{'bar':{'baz':...}}}.
  * The token location is preserved so that error messages refer to the original content being parsed.
+ * This parser can output duplicate keys, but that is fine given that it's used for document parsing. The mapping
+ * lookups will return the same mapper/field type, and we never load incoming documents in a map where duplicate
+ * keys would end up overriding each other.
```

Comment:
I wonder, if we're saying that this should *only* be used for document parsing maybe it ought to be a package-private class in `core.index.mapper` rather than in the XContent library?

Reply:
I missed it was part of xcontent, yes I think we should move it out of there. This is super specific and even dangerous to use besides where we use it now, I am thinking.